lib/utils/remote-build.ts

```diff
@@ -0,0 +1,219 @@
+	}).return();
+}
+
+async function handleBuilderMetadata(obj: BuilderMessage, build: RemoteBuild) {
+	switch (obj.resource) {
+		case 'cursor':
+			const readline = await import('readline');
+
+			if (obj.value == null) {
+				return;
+			}
+
+			const match = obj.value.match(CURSOR_METADATA_REGEX);
+
+			if (!match) {
+				// FIXME: Make this error nicer.
+				console.log(
+					'Warning: ignoring unknown builder command. You may experience odd build output. ' +
+						'Maybe you need to update resin-cli?',
+				);
+				return;
+			}
+
+			const value = match[1];
+			const amount = match[2] || 1;
+
+			switch (value) {
+				case 'erase':
+					readline.clearLine(process.stdout, 0);
+					process.stdout.write('\r');
+					break;
+				case 'up':
+					readline.moveCursor(process.stdout, 0, -amount);
+					break;
+				case 'down':
+					readline.moveCursor(process.stdout, 0, amount);
+					break;
+			}
+
+			break;
+		case 'buildLogId':
+			// The name of this resource is slightly dated, but this is the release
+			// id from the API. We need to save this so that if the user ctrl+c's the
+			// build we can cancel it on the API.
+			build.releaseId = parseInt(obj.value!, 10);
+			break;
+	}
+}
+
+function getBuilderMessageHandler(
+	build: RemoteBuild,
+): (obj: BuilderMessage) => Promise<void> {
+	return async (obj: BuilderMessage) => {
+		if (DEBUG_MODE) {
+			console.log(`[debug] handling message: ${JSON.stringify(obj)}`);
+		}
+		if (obj.type != null && obj.type === 'metadata') {
+			return handleBuilderMetadata(obj, build);
+		}
+		if (obj.message) {
+			const readline = await import('readline');
+			readline.clearLine(process.stdout, 0);
+
+			const message = obj.message.replace(TRIM_REGEX, '');
+			if (obj.replace) {
+				process.stdout.write(`\r${message}`);
+			} else {
+				console.log(message);
+			}
+		}
+	};
+}
+
+async function cancelBuildIfNecessary(build: RemoteBuild): Promise<void> {
+	if (build.releaseId != null) {
+		await build.sdk.pine.patch({
+			resource: 'release',
+			id: build.releaseId,
+			body: {
+				status: 'cancelled',
+				end_timestamp: Date.now(),
+			},
+		});
+	}
+}
+
+async function getRequestStream(build: RemoteBuild)
+	: Promise<{ stream: Stream.Duplex, uploadSpinner: { stop: () => void } }> {
+
+	const path = await import('path');
+	const visuals = await import('resin-cli-visuals');
+
+	const tarSpinner = new visuals.Spinner('Packaging the project source...');
+	tarSpinner.start();
+	// Tar the directory so that we can send it to the builder
+	const tarStream = await tarDirectory(path.resolve(build.source));
+	tarSpinner.stop();
+
+	if (DEBUG_MODE) {
+		console.log('[debug] Opening builder connection');
+	}
+	const post = request.post({
+		url: await getBuilderEndpoint(build.baseUrl, build.owner, build.app),
+		auth: {
+			bearer: build.auth,
+		},
+	});
+
+	const uploadSpinner = new visuals.Spinner('Uploading source package to resin cloud');
+	uploadSpinner.start();
+
+	tarStream.pipe(post);
```

Comment:
In fact, imo we should add a `data` handler here that's tied to `uploadSpinner.stop()`, and not return the spinner at all. I think that should work fine, and lets the code everywhere else ignore the spinners entirely.